tests/integration/cocoon/test_cocoon_portfolio_groups.py

```diff
@@ -0,0 +1,324 @@
+        ------------
+        Here we test attempting to add a portfolio which does not exist to a portfolio group.
+        """
+
+        test_df = pd.read_csv(
+            Path(__file__).parent.joinpath(
+                "data/port_group_tests/portfolio_group_portfolio_not_exist.csv"
+            )
+        )
+
+        responses = cocoon.cocoon.load_from_data_frame(
+            api_factory=self.api_factory,
+            scope=self.scope,
+            data_frame=test_df,
+            mapping_required={
+                "code": "PortGroupCode",
+                "display_name": "PortGroupDisplayName",
+            },
+            mapping_optional={"values.scope": "Scope", "values.code": "FundCode",},
+            file_type="portfolio_group",
+            property_columns=[],
+            properties_scope=None,
+        )
+
+        if responses.get("portfolio_groups").get("errors") != []:
+            for error in responses["portfolio_groups"]["errors"]:
+                print(json.loads(error.body)["title"])
+
+        self.assertEqual(
+            json.loads(responses["portfolio_groups"]["errors"][0].body)["name"],
+            "PortfolioNotFound",
+        )
+
+    def test_duplicate_portfolio(self):
+
+        """
+        Description:
+        ------------
+        Here we test attempting to add two of the same portfolios to a portfolio group.
+        """
+
+        duplicate_df = pd.read_csv(
+            Path(__file__).parent.joinpath(
+                "data/port_group_tests/portfolio_group_duplicate_portfolio.csv"
+            )
+        )
+
+        responses = cocoon.cocoon.load_from_data_frame(
+            api_factory=self.api_factory,
+            scope=self.scope,
+            data_frame=duplicate_df,
+            mapping_required={
+                "code": "PortGroupCode",
+                "display_name": "PortGroupDisplayName",
+            },
+            mapping_optional={"values.scope": "Scope", "values.code": "FundCode",},
+            file_type="portfolio_group",
+            property_columns=[],
+            properties_scope=None,
+        )
+
+        if responses.get("portfolio_groups").get("errors") != []:
+            for error in responses["portfolio_groups"]["errors"]:
+                print(json.loads(error.body)["title"])
+
+        duplicate_df["unique_combinations_for_test"] = (
+            duplicate_df["PortGroupCode"]
+            + "-"
+            + duplicate_df["FundCode"]
+            + "-"
+            + duplicate_df["Scope"]
+        )
+
+        self.assertEqual(
+            len(
+                [
+                    port_group
+                    for nested_group in [
+                        port_group.portfolios
+                        for port_group in responses["portfolio_groups"]["success"]
+                    ]
+                    for port_group in nested_group
+                ]
+            ),
+            len(duplicate_df["unique_combinations_for_test"].unique()),
+        )
```

Comment:
Am I right in saying that if you provide a duplicate in your request LUSID handles this gracefully and doesn't throw an error? I didn't see anything in our code to eliminate duplicates in the event of a creation request unlike where you take the set during an update.

Reply:
For duplicates:
CreatePortGroup - no errors. The latter request will be an "upsert" on the former request. 
AddPortfolioToGroup - this will throw errors if you provide duplicate portfolios and therefore we take the "set"